### DISCUSSION TITLE
Efficiency of `lax.cond` vs `jnp.where`

### DISCUSSION THREAD

**Discussion by rkruegs123** [jax-ml/jax#12281] [Q&A]
I have some code that is very conditional-heavy. However, it is taking very long to compile and/or run. At the moment, I implement all conditionals with `jnp.where`. I understand that there is an alternative, `lax.cond`, which (i) compiles to a different primitive, and (ii) doesn't execute both branches.

Can anyone speak to the differences between these two options when compiling to XLA, and more generally? Another point of discussion would be which maintains the "quality" of gradients better -- i.e. which gives cleaner gradients. Note that I am only using numerical conditions so both are a viable option for me.

**Comment by jakevdp** [comment]
The main difference, as you say, is that `jnp.where` (which lowers to `lax.select`) will evaluate both branches, while `lax.cond` will only evaluate a single branch. Note, however, that under a `vmap` transformation, `lax.cond` will be converted to `lax.select`, because the `cond` primitive only supports a single condition. I don't believe there should be any distinction in the quality of gradients generated by the two approaches.

**Reply by rkruegs123** [reply]
Thank you for your quick reply!

That makes sense regarding `vmap` and the gradients. What about compilation -- would you expect one to incur a larger compilation cost than the other? I know it may sound minor, but I have reason to believe that the part of my code containing many conditionals is causing slowdowns, so I am looking for ways to optimize them.
